Adding blobbers to newAllocation and removing prefferedBlobbers

moldis opened the pull request:
- removing PreferredBlobbers from NewAllocationRequest
- removing the functionality to fetch all blobbers from MPT and then find appropriate blobber to handle allocation
- instead the list of blobbers is to be expected in request and validated further
- freeAllocation is also updated with the same, it has now a blobbers key which take []string as input key

moldis committed:
:hammer: init commit

moldis committed:
fixup! :hammer: init commit

moldis committed:
:hammer: resolvig failed unit tests

moldis committed:
:art: removing old tests

moldis committed:
:hammer: fixing getallocationminlock

cnlangzi commented:
@moldis  don't forget to remove `prefferedBlobbers` stuff from gosdk and zbox

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -333,14 +325,9 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
 	blobberNodes, bSize, err := sc.selectBlobbers(
-		t.CreationDate, *allBlobbersList, sa, seed, balances)
+		t.CreationDate, *allBlobbersList, sa, balances, request.Blobbers)
+
```
We should not need to pass the `allBlobberList` here. 

We should expect hundreds of thousands of blobbers here. Use of  `allBlobbers` objects is deprecated and will be removed as part of optimisation.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/handler.go:
```diff
@@ -98,7 +98,8 @@ func (ssc *StorageSmartContract) GetAllocationMinLockHandler(ctx context.Context
 	blobberNodes, bSize, err := ssc.selectBlobbers(
-		creationDate, *allBlobbersList, sa, int64(creationDate), balances)
+		creationDate, *allBlobbersList, sa, balances, request.Blobbers)
+
```
We do not need the `allBlobbersList` here.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -426,7 +413,7 @@ func (sc *StorageSmartContract) selectBlobbers(
 	// size of allocation for a blobber
 	var bSize = (sa.Size + int64(size-1)) / int64(size)
 	var list = sa.filterBlobbers(allBlobbersList.Nodes.copy(), creationDate,
-		bSize, filterHealthyBlobbers(creationDate),
```
We should not pass the `allBlobberList` here. Instead, pass the `blobbers` input parameter.

Previously the `filterBlobbers` method iterated through `all` the blobbers filtering out the ones that did not fit the new allocation request parameters. 

We want to repurpose `filterBlobbers` to filter out any input blobbers that do not fit the allocation request parameters.

If any blobbers are filtered then we should expect a not enough blobbers error. 

https://github.com/0chain/0chain/issues/540 Uses complain that the error messaging here is unhelpful. Could we change `filterBlobbers` to be more informative about filtered out blobbers? 

One solution would be to return a list of filtered out blobber ids with the reason they were filtered. This could be appended to an error message if it turns out that there are not enough valid blobbers to form an allocation.

moldis replied:
@Sriep Maybe I misunderstood. My idea is to pass the list of blobbers (urls of blobbers) in the allocation endpoint, then this filter is picking up blobbers by URL and a few other filters like healthy, size etc. In any case, I will need to iterate through blobbers to get Blobber object?

Sriep replied:
A list of 100,000+ blobbers is too much to keep in the MPT. Takes over 50ms to get and update. So that is being removed soon. So pretend the allBlobbers list is not there, but leave the existing code, it will be removed shortly.

Instead, the list of blobbers checked here has to be loaded individually. Updating 40 items from the MPT is bad, but probably only about 10-20ms so we can handle it. 

I can's see any reason the same `filterBlobbers` method can't be used to check the user-selected list of blobbers the same way it filtered the all blobbers list. But could be wrong.

We want to avoid doing anything that potentially involves manipulating hundred of thousands of blobbers.

Sriep replied:
Actually, if you could remove the all blobbers list entirely from new allocation request, that would help.

Sriep replied:
I just modified https://github.com/0chain/0chain/pull/691 to supply `/getBlobbers` with information from the eventDb. I believe that the [all blobbers list](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/blobber.go#L17) is only used in new allocation requests and [/getBlobbes](https://github.com/0chain/0chain/blob/blobber-events/code/go/0chain.net/smartcontract/storagesc/handler.go#L53). The other references are just maintaining the list. 

Any endpoint that maintains the all blobbers list has to take >50ms with 100,000+ blobbers so removing it will be a solid step towards making our smartcontracts scalable.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation_test.go:
```diff
@@ -189,8 +186,9 @@ func TestSelectBlobbers(t *testing.T) {
+			var bl []string
 			outBlobbers, outSize, err := ssc.selectBlobbers(
-				now, blobbers, &sa, randomSeed, balances,
+				now, blobbers, &sa, balances, bl,
```
remove `blobbers`

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -172,7 +171,6 @@ func BenchmarkTests(
 					Expiration:                 common.Timestamp(viper.GetDuration(bk.StorageMinAllocDuration).Seconds()) + now,
 					Owner:                      data.Clients[0],
 					OwnerPublicKey:             data.PublicKeys[0],
-					PreferredBlobbers:          blobberUrls[:8],
```
Could rename this to the new Blobbers field rather than lose completely, otherwise I would expect a not enough blobbers error. 
Also, rename the test to 
```go
name:     "storage.new_allocation_request
```

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -140,7 +140,6 @@ func BenchmarkTests(
 					Expiration:                 common.Timestamp(viper.GetDuration(bk.StorageMinAllocDuration).Seconds()) + now,
 					Owner:                      data.Clients[0],
 					OwnerPublicKey:             data.PublicKeys[0],
-					PreferredBlobbers:          []string{},
```
We do not need this test anymore so can be removed.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_rest_tests.go:
```diff
@@ -103,7 +103,6 @@ func BenchmarkRestTests(
 					Expiration:                 2*common.Timestamp(viper.GetDuration(bk.StorageMinAllocDuration).Seconds()) + now,
 					Owner:                      data.Clients[0],
 					OwnerPublicKey:             data.PublicKeys[0],
-					PreferredBlobbers:          []string{},
```
We should probably replace the line with `Blobbers:          blobberUrls[:8],` rather than delete it. Otherwise, we should expect a not enough blobbers error.

Sriep reviewed (changes requested):
I think we need a  `max_blobbers_per_allocation` `sc.yaml` setting. Can't have users asking for, and getting 80,000 blobbers. This could default to 40.

moldis committed:
:art: removing allBlobbers

moldis committed:
:bug: resolving build

moldis committed:
:hammer: max available blobbers

moldis committed:
fixup! :hammer: max available blobbers

moldis committed:
:art: updating blobbers

moldis committed:
fixup! :art: updating blobbers

moldis committed:
:art: fixing unit tests

moldis committed:
:bug: minor fox

moldis committed:
:art: removing the commented code

moldis committed:
:bug: fixing unit tests

moldis committed:
:bug: fixing broken unit-tests

moldis committed:
:bug: cleaning up comments

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -119,36 +119,6 @@ func BenchmarkTests(
-			}(),
-		},
 		{
 			name:     "storage.new_allocation_request_preferred",
```
`storage.new_allocation`,  remove request_preferred.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -172,7 +142,6 @@ func BenchmarkTests(
 					Owner:                      data.Clients[0],
 					OwnerPublicKey:             data.PublicKeys[0],
-					PreferredBlobbers:          blobberUrls[:8],
 					ReadPriceRange:             PriceRange{0, state.Balance(viper.GetInt64(bk.StorageMaxReadPrice) * 1e10)},
```
Why no `newAllocationRequest.Blobbers = blobberUrls[0:4]` or similar  field? Should this not return a not enough blobbers error?

ssardana96 replied:
outdated

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -172,7 +142,6 @@ func BenchmarkTests(
-					PreferredBlobbers:          blobberUrls[:8],
 					ReadPriceRange:             PriceRange{0, state.Balance(viper.GetInt64(bk.StorageMaxReadPrice) * 1e10)},
 					WritePriceRange:            PriceRange{0, state.Balance(viper.GetInt64(bk.StorageMaxWritePrice) * 1e10)},
 					MaxChallengeCompletionTime: viper.GetDuration(bk.StorageMaxChallengeCompletionTime),
```
Remove `torage.new_allocation_request_diversify` benchmark below. We don't need it any more.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/blobber_test.go:
```diff
@@ -215,19 +212,15 @@ func Test_flow_reward(t *testing.T) {
+	var b1 *Client = blobs[0]
 
-	require.EqualValues(t, 202546280, alloc.restMinLockDemand())
+	// QUESTION HERE
```
The blobber objects used int `newAllocationRequestInternal`  don't have a valid Terms.MinLockDemand value. Get blobber from blockchain to make sure.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/blobber_test.go:
```diff
@@ -286,18 +283,21 @@ func Test_flow_reward(t *testing.T) {
-		require.EqualValues(t, 192418966, alloc.restMinLockDemand())
+
+		// TODO not clear for me
+		//require.EqualValues(t, 192418966, alloc.restMinLockDemand())
```
As above blobbers do not have valid Terms.MinLockDemand. Use blobbers from MTP not dodgy user input.

ssardana96 replied:
@Sriep didn't get it
I'm taking now blobbers from MPT only but, still MinLockDemand is 0
as all blobbers are getting entered in allAlloc() func, where blobbers are created for test and pushed to MPT

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -171,11 +170,11 @@ type newAllocationRequest struct {
 	WritePriceRange            PriceRange       `json:"write_price_range"`
 	MaxChallengeCompletionTime time.Duration    `json:"max_challenge_completion_time"`
 	DiversifyBlobbers          bool             `json:"diversify_blobbers"`
+	Blobbers				   []*StorageNode	`json:"blobbers"`
```
[]String = list of blobber ids should be good enough. 
Its painful getting all the blobbers from the MPT, but we can't assume users will enter valid data.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -336,20 +322,21 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
-	if seed, err = strconv.ParseInt(t.Hash[0:8], 16, 64); err != nil {
-		return "", common.NewError("allocation_creation_failed",
-			"Failed to create seed for randomizeNodes")
+	if len(request.Blobbers) > conf.MaxBlobbersPerAllocation{
```
We should probably check len(request.Blobbers) > request.DataShards + request.ParityShards as well, somewhere around here.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -336,20 +322,21 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
 
 	blobberNodes, bSize, err := sc.selectBlobbers(
-		t.CreationDate, *allBlobbersList, sa, seed, balances)
+		t.CreationDate, sa, balances, request.Blobbers)
```
We should get the blobbers from the MPT rather than rely on user input to be valid. 

Its painful but we  have `max_blobbers_per_allocation` to ensure that its ok. Plus its been budgeted for, check spreadsheet https://docs.google.com/spreadsheets/d/1oqY-gyt5itAJLxsh748CkFB8OOQsR8KClbzTrUfDf2A/edit#gid=677534893. Note spreadsheet assumes upcoming changes including partitions.


I suspect this is why you are having problems with the blobber_test.go minLockDemand results. The blobber.Terms.MinLockDemand field is zero in the input so givers wrong value on line 353 below.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/free_allocation.go:
```diff
@@ -30,8 +30,9 @@ func (frm *freeStorageMarker) decode(b []byte) error {
-	Marker             string `json:"marker"`
+	RecipientPublicKey string         `json:"recipient_public_key"`
+	Marker             string         `json:"marker"`
+	Blobbers           []*StorageNode `json:"blobbers"`
```
As for allocation. I think []string = list of blobber ids is all that's needed.

ssardana08 committed:
- validation of input blobber size
- removed storage.new_allocation_request_preferred

ssardana08 committed:
- input blobbers are taken from MPT
- unit tests corrected
- blobbers key added as []string for input blobbers by user

ssardana08 committed:
- corrected failed unit tests

ssardana08 committed:
- corrected failed unit tests

ssardana08 committed:
- made async flow to get blobbers

ssardana08 committed:
- making sequential calls to fetch blobber

ssardana08 committed:
- parallel get blobbers
- unit tests handled

ssardana96 commented:
@Sriep can you review it

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -384,6 +459,8 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
 	}
 
+	var allBlobbersList *StorageNodes
+	allBlobbersList, err = sc.getBlobbersList(balances)
```
Should be ok to remove this block of code.
```go
	var allBlobbersList *StorageNodes
	allBlobbersList, err = sc.getBlobbersList(balances)
	err = updateBlobbersInAll(allBlobbersList, blobberNodes, balances)
	if err != nil {
		return "", common.NewError("allocation_creation_failed", err.Error())
	}
```
The all blobbers list is depreciated and should be safe to remove. Currently very expensive and does nothing.

ssardana96 replied:
@Sriep couple of unit_tests again start to break because of this.
for now can we keep this? maybe we can take this in some other issues.

Sriep reviewed (changes requested):
Looks ok. 
The `zbox` and event database changes could allow system tests to pass, the exception is preferred blobbers - this might need renaming or changing in the sytem_tests.

`zbox newallocaion` remaining unchanged from a black-box perspective.

ssardana08 committed:
- error improved
- benchamrks were failing, corrected

dabasov commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -283,6 +283,59 @@ func (sc *StorageSmartContract) filterBlobbersByFreeSpace(now common.Timestamp,
+	output := make(chan *StorageNode, len(inputBlobberMap))
+	defer close(output)
+	for pool := 0; pool < poolSize; pool++ {
+		go func() {
```
I don't think we should use concurrency in smart contracts tbh

dabasov replied:
do we get any performance boost?

Sriep replied:
My benchmarks suggest it can halve the speed of smart contracts. So using parallel for loops could well allow us to have a maximum of say 40 blobbers per allocation rather than 20.

What is the problem?

ssardana08 replied:
40 blobbers, concurrency:8 get_blobbers	{"time_taken": "167ns"}
40 blobbers, concurrency:4 get_blobbers	{"time_taken": "208ns"}
40 blobbers, concurrency disabled get_blobbers	{"time_taken": "250ns"}

ssardana08 replied:
these are timings for this function with and without cocurrency

dabasov replied:
afaik we never used it and all structures used in smartcontracts don't have proper synchronization, if we allow to use it somewhere we should code accordingly, for example StateContextI is not designed for concurrent usage, look here [field access](https://github.com/0chain/0chain/blob/c73e2717354d70f33a33fe1a73487ca85f506d14/code/go/0chain.net/chaincore/chain/state/state_context.go#L290)

dabasov replied:
in my opinion we should parallelize work on some higher level and implement smartcontracts just as not concurrent code

ssardana08 committed:
- message corrected

peterlimg commented:
Please fix the conflicts, thanks

Sriep committed:
error messages

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/handler.go:
```diff
@@ -254,23 +256,20 @@ func (ssc *StorageSmartContract) GetAllocationMinLockHandler(ctx context.Context
 	if err = request.decode([]byte(allocData)); err != nil {
-		return "", common.NewErrInternal("can't decode allocation request", err.Error())
+		logging.Logger.Error("can't decode allocation request",
+			zap.Error(err))
```
Return error instead of logging only

ssardana08 commented:
@samtholiya is this completed? @dabasov ?
if not needed, can we close it?

samtholiya commented:
This is not needed
